### PR TITLE
validate: Expose hooks to inject custom behavior during traversal 

### PR DESCRIPTION
**Pull request by alexzielenski**
Adds a new Option to the schema validator for users to provider custom constructors for subfield and subindex-validators created during traversal.

/cc @stts @apelisse @jpbetz

### CODE REVIEW

File: pkg/validation/validate/schema.go [hunk 1]
```diff
@@ -78,6 +78,12 @@ func NewSchemaValidator(schema *spec.Schema, rootSchema interface{}, root string
 	for _, o := range options {
 		o(&s.Options)
 	}
+
+	if s.Options.subIndexValidator == nil {
+		s.Options.subPropertyValidator = s.SubPropertyValidator
```
Comment (sttts) [inline pkg/validation/validate/schema.go:83]: why override both after checking only one?

File: pkg/validation/validate/schema.go [hunk 2]
```diff
@@ -91,6 +97,14 @@ func NewSchemaValidator(schema *spec.Schema, rootSchema interface{}, root string
 	return &s
 }
 
+func (s *SchemaValidator) SubPropertyValidator(field string, sch *spec.Schema) valueValidator {
```
Comment (sttts) [inline pkg/validation/validate/schema.go:100]: for consistency let's call these `New*`

File: pkg/validation/validate/slice_validator.go [hunk 3]
```diff
@@ -54,7 +54,7 @@ func (s *schemaSliceValidator) Validate(data interface{}) *Result {
 	size := val.Len()
 
 	if s.Items != nil && s.Items.Schema != nil {
-		validator := NewSchemaValidator(s.Items.Schema, s.Root, s.Path, s.KnownFormats, s.Options.Options()...)
+		validator := s.Options.subIndexValidator(0, s.Items.Schema)
```
Comment (sttts) [inline pkg/validation/validate/slice_validator.go:57]: this is wrong, will give you the wrong index in results

Comment (sttts) [reply pkg/validation/validate/slice_validator.go:57]: actually not sure, it is wrong. But least it is surprising to pass zero. Why the index param in the first place?

File: pkg/validation/validate/ratcheting.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	s := NewSchemaValidator(r.Schema, r.Root, r.Path, r.KnownFormats, opts...)
+
+	res := s.Validate(new)
+
+	if res.IsValid() {
+		return res
+	}
+
+	if reflect.DeepEqual(old, new) {
+		//!TODO: only consider errors with paths on "."
```
Comment (sttts) [inline pkg/validation/validate/ratcheting.go:43]: yes, without filtering by `.` we will deepequal all the way up to the route. That's heavy and unnecessary.

File: pkg/validation/validate/schema.go (hunk 2 above)
Comment (alexzielenski) [reply pkg/validation/validate/schema.go]: renamed the option fields

File: pkg/validation/validate/schema.go (hunk 1 above)
Comment (alexzielenski) [reply pkg/validation/validate/schema.go]: fixed

**Comment by apelisse** [comment]
Looks good to me, Stefan, can you tag it?
/approve

**Comment by jpbetz** [comment]
/approve

I'll defer to @sttts for lgtm

File: pkg/validation/spec/info.go [hunk 5]
```diff
@@ -53,6 +53,10 @@ func (e Extensions) GetBool(key string) (bool, bool) {
 // GetStringSlice gets a string value from the extensions
 func (e Extensions) GetStringSlice(key string) ([]string, bool) {
 	if v, ok := e[strings.ToLower(key)]; ok {
+		if alreadyStringSlice, ok := v.([]string); ok {
```
Comment (sttts) [inline pkg/validation/spec/info.go:56]: xref https://github.com/kubernetes/kube-openapi/pull/413#issuecomment-1636119257

**Comment by sttts** [comment]
Lgtm with exception of https://github.com/kubernetes/kube-openapi/pull/406/files#r1263940329.

File: pkg/validation/spec/info.go (hunk 5 above)
Comment (alexzielenski) [reply pkg/validation/spec/info.go]: removed this bit

**Comment by alexzielenski** [comment]
removed string slice logic

**Pull request merged**